### ISSUE TITLE
Publishing occasionally fails with struct unpack error

### ISSUE THREAD

**Issue by Alberts00** [eclipse/paho.mqtt.python#255]
Hello,
I'm occasionally seeing an issue whilst publishing a message and was able to catch the exception in a debugger after a long time period.

The debugger output when the exception occurred can be seen below.
![2017-12-15-20 30 55](https://user-images.githubusercontent.com/1869097/34055838-9cfd7816-e1d9-11e7-8d5d-d30ece36453d.png)

I do not think this is a network related issue since it happens on publish before any packet is being sent out.
![2017-12-15-20 33 31](https://user-images.githubusercontent.com/1869097/34055867-afb4ab5a-e1d9-11e7-8cc5-dc6643cffa01.png)
![2017-12-15-20 38 45](https://user-images.githubusercontent.com/1869097/34055869-b60ef62c-e1d9-11e7-90f8-eaa3e36ac2e3.png)

Full data at the time of _handle_publish, client.py:2466 can be seen [here](https://gist.github.com/Alberts00/5bfa1ef5a10fa297ecc8f2285289dbc5).

Full traceback:
```

 Exception in thread Thread-6:
Traceback (most recent call last):
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 2650, in _thread_main
    self.loop_forever(retry_first_connection=True)
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 1481, in loop_forever
    rc = self.loop(timeout, max_packets)
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 1003, in loop
    rc = self.loop_read(max_packets)
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 1284, in loop_read
    rc = self._packet_read()
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 1849, in _packet_read
    rc = self._packet_handle()
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 2305, in _packet_handle
    return self._handle_publish()
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 2466, in _handle_publish
    (topic, packet) = struct.unpack(pack_format, packet)
struct.error: bad char in struct format

Traceback (most recent call last):
  File "/media/storage/wdir/pycharm/helpers/pydev/pydevd.py", line 1668, in <module>
    main()
  File "/media/storage/wdir/pycharm/helpers/pydev/pydevd.py", line 1662, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/media/storage/wdir/pycharm/helpers/pydev/pydevd.py", line 1072, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/media/storage/wdir/pycharm/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/media/storage/priv/projects/openhab_ss/consume.py", line 92, in <module>
    read_serial(ser, client)
  File "/media/storage/priv/projects/openhab_ss/consume.py", line 84, in read_serial
    client.loop() # To make sure it reconnects if connection is dropped
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 1003, in loop
    rc = self.loop_read(max_packets)
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 1284, in loop_read
    rc = self._packet_read()
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 1849, in _packet_read
    rc = self._packet_handle()
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 2305, in _packet_handle
    return self._handle_publish()
  File "/home/alberts00/.pyenv/versions/3.6.2/lib/python3.6/site-packages/paho/mqtt/client.py", line 2466, in _handle_publish
    (topic, packet) = struct.unpack(pack_format, packet)
struct.error: bad char in struct format
```

Hardware and software used:
Kernel: 3.19.0-73-generic #81~14.04.1-Ubuntu
Data: Input from serial directly passed onto paho-mqtt
Python version: 3.6 (via pyenv), however 3.5 (via Ubuntu repository) is also experiencing same issue.
Python package versions:
- coverage (4.4.2)
- paho-mqtt (1.3.1)
- pip (9.0.1)
- pyserial (3.4)
- setuptools (28.8.0)

I will be very thankful for any suggestions on what to debug further were provided.
Let me know if any additional information is required.

Thanks for taking your time looking through this issue, have a great day!

**Comment by PierreF** [comment]
Hi, thanks for the detailed error reporting.

Your traceback make me think you are calling loop() and loop_forever() from 2 thread and that probably the cause of your error, loop() is not thread-safe.

**Comment by phretor** [comment]
@PierreF not safe even if each thread has its own `Client` object?

**Comment by Alberts00** [comment]
Hey @PierreF,
Thanks for pointing that out, I've realised that was indeed the case.
I was running client.loop_start() inside on_connect() and running client.loop() after data has been fetched from serial port (inside Main thread).
Unfortunately I did not notice anywhere in the documentation that both of those functions are not thread safe (might be wise to document this).
If no action regarding the documentation issue will be taken care of I presume this issue can be closed. Thanks for taking your time looking into this!
